### ISSUE TITLE
Doesn't work anymore with Remix `1.12.0`

### ISSUE THREAD

**Issue by kwiat1990** [jrestall/remix-stubs#10]
After updating Remix to the version `1.12.0` the stub doesn't work anymore as the interfaces have changed and TS is complaing about some type `'{ actionData: RouteData | undefined; appState: { trackBoundaries: boolean; trackCatchBoundaries: boolean; catchBoundaryRouteId: null; renderBoundaryRouteId: null; loaderBoundaryRouteId: null; error: undefined; catch: undefined; }; matches: { ...; }[]; routeData: RouteData; manifest: AssetsManifest; routeModules: Rou...' is not assignable to type 'EntryContext'.
  Object literal may only specify known properties, and 'actionData' does not exist in type 'EntryContext'` used in `createRemixContext` function.

Have you already tried the unstable `@remix-run/testing?` package?

**Comment by jrestall** [comment]
Hi @kwiat1990, for version 1.12.0+ I'd recommend moving to @remix-run/testing since that's the go forward approach anyway. It's been working great for me across 100+ unit tests.

**Comment by kwiat1990** [comment]
My Setup relies on Vitest and remix stub approach. After I did update Vitest to `0.28.4` and switch to `remix-run/testing` package all tests, which depends on initial loader or action are broken as those values stay `undefined`. For example:

```ts
const RemixStub = createRemixStub([
  {
    path: "/",
    element: <TestSubject />,
    loader() {
      return { theme: Theme.LIGHT };
    },
  },
]);

function TestSubject() {
  const data = useLoaderData(); => data is undefined
  return (
    <ThemeProvider specifiedTheme={data.theme}>
      <Header />
    </ThemeProvider>
  );
}

// ...

it("has all elements", () => {
    render(<RemixStub hydrationData={{ loaderData: { "/": { theme: Theme.LIGHT } } }} />);
    expect(screen.getByRole("navigation")).toBeVisible();
  });
```

**Comment by jrestall** [comment]
There's a change in behavior with @remix-run/testing where the loaderData is now keyed on the React Router IDs by default e.g. 0,1,2,1:2 rather than the path. So your example above I think would be `loaderData: { "0": { theme: Theme.LIGHT } } }}`.

However you should specify a route id and use that to key the loader data instead as its clearer. e.g.

```
const RemixStub = createRemixStub([
  {
    id: "test",
    path: "/",
    element: <TestSubject />,
    loader() {
      return { theme: Theme.LIGHT };
    },
  },
]);

it("has all elements", () => {
    render(<RemixStub hydrationData={{ loaderData: { "test": { theme: Theme.LIGHT } } }} />);
    expect(screen.getByRole("navigation")).toBeVisible();
  });
```

**Comment by kwiat1990** [comment]
Thanks for the example! I have seen yesterday a thread from Kent C. Dodds in Remix repository discussions but I didn't get the idea how the IDs should be set in the first place. With this small adjustment it all works now.

**Comment by nobleach** [comment]
I know that neither of you are a "support person" for this but I've run into the same experience. Even with the updates to `createRemixStub` inside `@remix-run/testing` I'm seeing that I cannot pass loader data to a component. The component basically "crashes" on `useLoaderData`. My setup is like this: 

```typescript
    const mockMenuLinksResponse = {
      data: {
        menuLinks: [
          {
            id: '1e98d9e9-e51a-46ec-b8a2-2299519065e8',
            name: 'how-it-works',
            label: 'How it Works',
          },
        ],
      },
    };

    let RemixStub = createRemixStub([
      {
        id: 'test',
        path: '/test-route',
        loader: () => mockMenuLinksResponse,
        element: <HowItWorksLink />,
      },
    ]);

        const { queryByText } = render(
      <RemixStub
        initialEntries={['/test-route']}
        hydrationData={{ loaderData: { test: mockMenuLinksResponse } }}
      />,
      store
    );

    screen.debug();
```
I'm trying to log inside my component:
```typescript
export default function HowItWorksLink() {
  console.log('#### at least I rendered');
  const {
    menuLinks: { data },
  } = useLoaderData();
  console.log('#### from useLoaderData', menuLinks);
// snip
```
The test blows up:
```sh
#### at least I rendered                                                                                                                                                                                     
<body>                                                                                                                                                                                                       
  <div>                                                                                                                                                                                                      
    <h2>                                                                                                                                                                                                     
      Unexpected Application Error!                                                                                                                                                                          
    </h2>                                                                                                                                                                                                    
    <h3     
   ...
    Error: Uncaught [TypeError: Cannot read properties of undefined (reading 'data')]                      
```

I've been reading the source of all the concerned pieces and I can't see any reason that loader isn't passing the data so that the component can use it via `useLoaderData`. It seems like it should be dead simple.

**Comment by jrestall** [comment]
Hi @nobleach, I think you will have the best support posting in the remix discord or discussions as I can't immediately see an issue with your code. `@remix-run/testing` is working very well for me so there should be a way to get this working.

I'd probably debug through the remix useLoaderData code to see if that helps determine why the loader data isn't there.

**Comment by kwiat1990** [comment]
@nobleach I think you have wrongly destructured the loader and that's why your test is broken. I managed to get it work by adjusting your code:

```tsx
// how-it-works.tsx

import { LoaderArgs } from "@remix-run/node";
import { useLoaderData } from "@remix-run/react";

export const loader = async ({ params, request }: LoaderArgs) => {
  return {
    data: {
      menuLinks: [
        {
          id: "1e98d9e9-e51a-46ec-b8a2-2299519065e8",
          name: "how-it-works",
          label: "How it Works",
        },
      ],
    },
  };
};

export default function HowItWorksLink() {
  const {
    data: { menuLinks },
  } = useLoaderData();

  console.log(menuLinks);

  return null;
}
```

```tsx
// how-it-works.test.tsx

import { unstable_createRemixStub as createRemixStub } from "@remix-run/testing";
import { render } from "@testing-library/react";
import HowItWorks from "./how-it-works";

const mockMenuLinksResponse = {
  data: {
    menuLinks: [
      {
        id: "1e98d9e9-e51a-46ec-b8a2-2299519065e8",
        name: "name from test data",
        label: "label from test data",
      },
    ],
  },
};

const RemixStub = createRemixStub([
  {
    id: "test",
    path: "/test-route",
    loader: () => mockMenuLinksResponse,
    element: <HowItWorks />,
  },
]);

describe("how-it-works", async () => {
  it("renders component", async () => {
    const { debug } = render(
      <RemixStub
        initialEntries={["/test-route"]}
        hydrationData={{ loaderData: { test: mockMenuLinksResponse } }}
      />
    );

    debug();
  });
});

**Comment by nobleach** [comment]
Thanks to both of you, and again, I apologize as I know this isn't a support forum. This is just the closest I've found to others experiencing my same problem and successfully solving it. I went back to this https://github.com/remix-run/remix/blob/dev/packages/remix-testing/__tests__/stub-test.tsx#L44 and realized, if it's working/passing for the Remix team, I had to be at fault. I've slowly been reworking my test from this and sure enough @kwiat1990 I was destructuring incorrectly. Thanks again.
